monitor/monitor.go

```diff
@@ -0,0 +1,141 @@
+package monitor
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"net/http"
+	"time"
+
+	ctClient "github.com/google/certificate-transparency-go/client"
+	"github.com/google/certificate-transparency-go/jsonclient"
+	"github.com/jmhodges/clock"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+// monitorStat is a struct collecting up various prometheus metrics a monitor
+// will export/track.
+type monitorStats struct {
+	sthTimestamp *prometheus.GaugeVec
+	sthFailures  *prometheus.CounterVec
+	sthLatency   *prometheus.HistogramVec
+}
+
+const (
+	// sthTimeout controls how long should each STH fetch wait before timing out
+	sthTimeout = time.Second * 15
+)
+
+var (
+	// internetFacingBuckets are histogram buckets suitable for measuring
+	// latencies that involve traversing the public internet.
+	internetFacingBuckets               = []float64{.1, .25, .5, 1, 2.5, 5, 7.5, 10, 15, 30, 45}
+	stats                 *monitorStats = &monitorStats{
+		sthTimestamp: promauto.NewGaugeVec(prometheus.GaugeOpts{
+			Name: "sth_timestamp",
+			Help: "Timestamp of observed CT log signed tree head (STH)",
+		}, []string{"uri"}),
+		sthFailures: promauto.NewCounterVec(prometheus.CounterOpts{
+			Name: "sth_failures",
+			Help: "Count of failures fetching CT log signed tree head (STH)",
+		}, []string{"uri"}),
+		sthLatency: promauto.NewHistogramVec(prometheus.HistogramOpts{
+			Name:    "sth_latency",
+			Help:    "Latency observing CT log signed tree head (STH)",
+			Buckets: internetFacingBuckets,
+		}, []string{"uri"}),
+	}
+)
+
+// Monitor is a struct for monitoring a CT log.
+type Monitor struct {
+	logger *log.Logger
+	clk    clock.Clock
+	stats  *monitorStats
+	logURI string
+	logKey string
+	// We use the read-only CheckLogClient interface here because the monitor
+	// never writes to the log and it allows easier mocking
+	client ctClient.CheckLogClient
```

Comment:
The above comment isn't true: we do intend to submit certificates (https://github.com/letsencrypt/ct-woodpecker/issues/2)

Reply:
Also note that best practice for mocking in tests is generally to define local interfaces with the subset of methods you care about.

Reply:
Fixed in b307f66